### ISSUE TITLE
Closured instance variable of a struct results in invalid memory access

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#7771]
A closured ivar in a struct can be accessed but when calling a method on it, it runs into a segfault:

```cr
struct Foo
  def initialize
    @foo = ["foo"]
    -> do
      @foo << "bar" # Error: Invalid memory access
    end.call
  end
end

Foo.new
```` 

This works however:
```cr
struct Foo
  def initialize
    @foo = ["foo"]
    -> do
      @foo # => ["foo"]
    end.call
  end
end

Foo.new
```

**Comment by RX14** [comment]
The access working is just a result of the access being optimized out as the value is never used.

I'm certain that this example cannot compile.

**Comment by straight-shoota** [comment]
Hm, I'm pretty sure I had an example where the value could actually be used... But it seems that's not the case. And your explanation makes sense.
When it can't work, closuring a struct ivar should be a compiler error.

**Comment by asterite** [comment]
I'll try to debug it today, but I don't think closuring structs work well at all.

**Comment by asterite** [comment]
Actually, it seems related to `initialize`...

**Comment by asterite** [comment]
Yeah, the problem is that the compiler detects that `self` is closured, so it allocates that in the heap and stores the struct there... when the `initialize` method begins. At that point everything is zero inside the struct. Then it's never changed in the closured data (because it's not a reference, like a class).

Try this:

```crystal
struct Foo
  def initialize
    @bar = 1
    ->do
      p @bar
    end
  end
end

Foo.new
```

I don't have a solution for this. Maybe structs shouldn't be allowed to be closured.

**Comment by asterite** [comment]
Actually, maybe it can be done, but it has to be specific to `initialize`: reading and assigning to instance vars there must write to the closured data (`self`). But at the end a copy of the struct will be returned... I guess that at long that it doesn't segfault it will be fine. I'll see if I can fix it somehow.

**Comment by asterite** [comment]
Regardless of the fix (which I might not do soon because the use case is a bit contrived): how did you stumble upon this? What were you trying to do?

**Comment by straight-shoota** [comment]
IIRC this was mentioned by someone else, I just wrote the issue. But I'm not sure right now .. usually I mention the source.

**Comment by oprypin** [comment]
I feel like this issue needs a bump, since #9311 was closed as duplicate of this

**Comment by Blacksmoke16** [comment]
This was encountered on the discord when using `at_exit` within `initialize`. E.g. something along the lines of this, with a `TCPServer` also running.

```cr
require "socket"

struct Client
  def initialize(host : String, port : Int32)
    @tcp_socket = TCPSocket.new host, port

    at_exit do
      @tcp_socket.send "foo"
    end
  end
end

client =Client.new "127.0.0.1", 1234
```

**Comment by ysbaddaden** [comment]
Here's a related repro (no prelude, no segfault):

```crystal
struct Foo
  @value = 1

  def bug
    fn = -> { @value = 2 }
    fn.call
    @value
  end
end

foo = Foo.new
foo.bug # => 1
```

The captured block silently operates on a copy of `self`.

It makes sense because a struct is passed by-value and there's no telling whether a pointer to `self` would be valid when the proc is called, but that's clearly not the expected behavior. Even if the closured block would only read from `self`, accessing a staled copy is still probably not expected.

So, I'm not sure this should compile.
